### ISSUE TITLE
Unexpected Behaviour with Model Checkpointing and val_check_interval (again)

### ISSUE THREAD

**Issue by jamarju** [PyTorchLightning/pytorch-lightning#8190]
## 🐛 Bug

Same as: #1764 

## Please reproduce using the BoringModel

I'm terribly sorry that I don't have the time for this but I hope this bug report still gets accepted as it is not too hard to reproduce :(

### To Reproduce

Use this version:

```
>>> pl.__version__
'1.3.7post0'
```

Use a model that improves `val_loss` as it trains.

I'm using `LightningCLI`. Relevant yaml config:

```
trainer:
  gpus: 0,1,2
  accelerator: ddp
  val_check_interval: 0.2
  callbacks:
    - class_path: pytorch_lightning.callbacks.ModelCheckpoint
      init_args:
        monitor: 'val_loss'
        verbose: true
        mode: 'min'
        filename: '04_{epoch}_{val_loss:.6f}'
        save_weights_only: true
```

See how after 0.50505 none of the subsequent **better** losses are recorded:

```
poch 0:  20%|████████▍                                 | 22/109 [00:06<00:27,  3.19it/s, loss=1.05, v_num=88, val_loss=1.030, lr_trf=7.38e-7, lr_clf=7.38e-7Epoch 0, global step 17: val_loss reached 1.03065 (best 1.03065), saving model to "/space/ml/commonlit/lightning_logs/version_88/checkpoints/04_epoch=0_val_loss=1.030650.ckpt" as top 1
Epoch 0:  40%|████████████████▌                        | 44/109 [00:14<00:20,  3.13it/s, loss=0.984, v_num=88, val_loss=0.961, lr_trf=1.78e-6, lr_clf=1.78e-6Epoch 0, global step 35: val_loss reached 0.96075 (best 0.96075), saving model to "/space/ml/commonlit/lightning_logs/version_88/checkpoints/04_epoch=0_val_loss=0.960752.ckpt" as top 1
Epoch 0:  61%|████████████████████████▊                | 66/109 [00:21<00:13,  3.10it/s, loss=0.954, v_num=88, val_loss=0.893, lr_trf=3.36e-6, lr_clf=3.36e-6Epoch 0, global step 53: val_loss reached 0.89280 (best 0.89280), saving model to "/space/ml/commonlit/lightning_logs/version_88/checkpoints/04_epoch=0_val_loss=0.892801.ckpt" as top 1
Epoch 0:  81%|█████████████████████████████████        | 88/109 [00:28<00:06,  3.09it/s, loss=0.724, v_num=88, val_loss=0.694, lr_trf=5.23e-6, lr_clf=5.23e-6Epoch 0, global step 71: val_loss reached 0.69430 (best 0.69430), saving model to "/space/ml/commonlit/lightning_logs/version_88/checkpoints/04_epoch=0_val_loss=0.694305.ckpt" as top 1
Epoch 0: 100%|████████████████████████████████████████| 109/109 [00:35<00:00,  3.06it/s, loss=0.597, v_num=88, val_loss=0.621, lr_trf=7.09e-6, lr_clf=7.09e-6Epoch 0, global step 89: val_loss reached 0.62113 (best 0.62113), saving model to "/space/ml/commonlit/lightning_logs/version_88/checkpoints/04_epoch=0_val_loss=0.621130.ckpt" as top 1
Epoch 1:  20%|████████▎                                | 22/109 [00:06<00:27,  3.20it/s, loss=0.537, v_num=88, val_loss=0.644, lr_trf=8.94e-6, lr_clf=8.94e-6Epoch 1, global step 111: val_loss was not in top 1
Epoch 1:  40%|████████████████▉                         | 44/109 [00:13<00:19,  3.29it/s, loss=0.52, v_num=88, val_loss=0.577, lr_trf=9.82e-6, lr_clf=9.82e-6Epoch 1, global step 129: val_loss reached 0.57692 (best 0.57692), saving model to "/space/ml/commonlit/lightning_logs/version_88/checkpoints/04_epoch=1_val_loss=0.576923.ckpt" as top 1
Epoch 1:  61%|████████████████████████▊                | 66/109 [00:20<00:13,  3.20it/s, loss=0.524, v_num=88, val_loss=0.621, lr_trf=9.99e-6, lr_clf=9.99e-6Epoch 1, global step 147: val_loss was not in top 1
Epoch 1:  81%|█████████████████████████████████        | 88/109 [00:27<00:06,  3.25it/s, loss=0.562, v_num=88, val_loss=0.594, lr_trf=9.88e-6, lr_clf=9.88e-6Epoch 1, global step 165: val_loss was not in top 1
Epoch 1: 100%|████████████████████████████████████████| 109/109 [00:33<00:00,  3.24it/s, loss=0.455, v_num=88, val_loss=0.545, lr_trf=9.62e-6, lr_clf=9.62e-6Epoch 1, global step 183: val_loss reached 0.54479 (best 0.54479), saving model to "/space/ml/commonlit/lightning_logs/version_88/checkpoints/04_epoch=1_val_loss=0.544790.ckpt" as top 1
Epoch 2:  20%|████████▎                                | 22/109 [00:06<00:27,  3.19it/s, loss=0.434, v_num=88, val_loss=0.592, lr_trf=9.13e-6, lr_clf=9.13e-6Epoch 2, global step 205: val_loss was not in top 1
Epoch 2:  40%|████████████████▌                        | 44/109 [00:13<00:19,  3.28it/s, loss=0.461, v_num=88, val_loss=0.531, lr_trf=8.59e-6, lr_clf=8.59e-6Epoch 2, global step 223: val_loss reached 0.53115 (best 0.53115), saving model to "/space/ml/commonlit/lightning_logs/version_88/checkpoints/04_epoch=2_val_loss=0.531149.ckpt" as top 1
Epoch 2:  61%|████████████████████████▊                | 66/109 [00:20<00:13,  3.20it/s, loss=0.389, v_num=88, val_loss=0.524, lr_trf=7.95e-6, lr_clf=7.95e-6Epoch 2, global step 241: val_loss reached 0.52437 (best 0.52437), saving model to "/space/ml/commonlit/lightning_logs/version_88/checkpoints/04_epoch=2_val_loss=0.524372.ckpt" as top 1
Epoch 2:  81%|█████████████████████████████████        | 88/109 [00:27<00:06,  3.17it/s, loss=0.392, v_num=88, val_loss=0.517, lr_trf=7.22e-6, lr_clf=7.22e-6Epoch 2, global step 259: val_loss was not in top 1
Epoch 2: 100%|████████████████████████████████████████| 109/109 [00:34<00:00,  3.18it/s, loss=0.437, v_num=88, val_loss=0.535, lr_trf=6.43e-6, lr_clf=6.43e-6Epoch 2, global step 277: val_loss was not in top 1
Epoch 3:  20%|████████▎                                | 22/109 [00:06<00:27,  3.19it/s, loss=0.317, v_num=88, val_loss=0.518, lr_trf=5.41e-6, lr_clf=5.41e-6Epoch 3, global step 299: val_loss reached 0.51775 (best 0.51775), saving model to "/space/ml/commonlit/lightning_logs/version_88/checkpoints/04_epoch=3_val_loss=0.517754.ckpt" as top 1
Epoch 3:  40%|████████████████▉                         | 44/109 [00:14<00:20,  3.12it/s, loss=0.36, v_num=88, val_loss=0.505, lr_trf=4.56e-6, lr_clf=4.56e-6Epoch 3, global step 317: val_loss reached 0.50505 (best 0.50505), saving model to "/space/ml/commonlit/lightning_logs/version_88/checkpoints/04_epoch=3_val_loss=0.505046.ckpt" as top 1
Epoch 3:  61%|████████████████████████▊                | 66/109 [00:21<00:13,  3.10it/s, loss=0.313, v_num=88, val_loss=0.520, lr_trf=3.73e-6, lr_clf=3.73e-6Epoch 3, global step 335: val_loss was not in top 1
Epoch 3:  81%|█████████████████████████████████        | 88/109 [00:27<00:06,  3.17it/s, loss=0.315, v_num=88, val_loss=0.517, lr_trf=2.93e-6, lr_clf=2.93e-6Epoch 3, global step 353: val_loss was not in top 1
Epoch 3: 100%|████████████████████████████████████████| 109/109 [00:34<00:00,  3.18it/s, loss=0.336, v_num=88, val_loss=0.516, lr_trf=2.19e-6, lr_clf=2.19e-6Epoch 3, global step 371: val_loss was not in top 1
Epoch 4:  20%|████████▎                                | 22/109 [00:06<00:27,  3.19it/s, loss=0.272, v_num=88, val_loss=0.504, lr_trf=1.39e-6, lr_clf=1.39e-6Epoch 4, global step 393: val_loss was not in top 1
Epoch 4:  40%|█████████████████▎                         | 44/109 [00:13<00:19,  3.28it/s, loss=0.262, v_num=88, val_loss=0.503, lr_trf=8.6e-7, lr_clf=8.6e-7Epoch 4, global step 411: val_loss was not in top 1
Epoch 4:  61%|████████████████████████▊                | 66/109 [00:19<00:12,  3.31it/s, loss=0.287, v_num=88, val_loss=0.504, lr_trf=4.45e-7, lr_clf=4.45e-7Epoch 4, global step 429: val_loss was not in top 1
Epoch 4:  81%|█████████████████████████████████        | 88/109 [00:26<00:06,  3.33it/s, loss=0.267, v_num=88, val_loss=0.504, lr_trf=1.62e-7, lr_clf=1.62e-7Epoch 4, global step 447: val_loss was not in top 1
Epoch 4: 100%|████████████████████████████████████████| 109/109 [00:32<00:00,  3.31it/s, loss=0.291, v_num=88, val_loss=0.503, lr_trf=1.81e-8, lr_clf=1.81e-8Epoch 4, global step 465: val_loss was not in top 1
Epoch 4: 100%|████████████████████████████████████████| 109/109 [00:34<00:00,  3.18it/s, loss=0.283, v_num=88, val_loss=0.503, lr_trf=8.07e-9, lr_clf=8.07e-9]Epoch 4, global step 469: val_loss was not in top 1
Epoch 4: 100%|████████████████████████████████████████| 109/109 [00:34<00:00,  3.18it/s, loss=0.283, v_num=88, val_loss=0.503, lr_trf=8.07e-9, lr_clf=8.07e-9]
```

### Expected behavior

Model should be checkpointed on `val_loss` < .50505 but it does not.

**Comment by jamarju** [comment]
Addendum: on a single GPU model checkpointing behaves properly. The above problem shows specifically on multi-GPU + ddp.

**Comment by awaelchli** [comment]
are you logging `val_loss` with `sync_dist=True`?

**Comment by jamarju** [comment]
No. I will try that and report back. Thank you.

**Comment by mortonjt** [comment]
I had similar issues when saving the checkpoints.  In one of the previous versions of pytorch-lightning ( I don't know which unfortunately, but is after the 1.0.0 release), there was new functionality added where checkpointing specifically checks the output stored by `self.log(metric, value)`.  If you don't use `self.log` within your validation_epoch, this will cause a silent failure and no intermediate checkpoints will be stored.  I have a hunch that is what is going on with @jamarju's problem.

@awaelchli this is not a bug, but it can be challenging to keep up with the new pytorch-lightning updates -- it would tremendously useful if warnings were thrown to at least alert users of these silent errors (debugging checkpoint callbacks can also be quite time consuming).

**Comment by jamarju** [comment]
> are you logging `val_loss` with `sync_dist=True`?

Hello Adrian, this fixed the issue in my case. Thank you. May I suggest that `self.log` produces a warning when used in a `ddp`.

**Comment by awaelchli** [comment]
Hi @jamarju , @mortonjt 
The warning is a nice idea, however note most of the time we don't want to log with sync-dist and such a warning could become annoying. I agree for checkpointing it would make sense to include the synced loss. However again, computing just the mean (default for sync-dist) would not always compute the correct value. It depends on the metric we are using. And this problem is what [torchmetrics](https://torchmetrics.readthedocs.io/en/latest/) is here to solve. 

What do you think?

**Comment by jamarju** [comment]
> And this problem is what [torchmetrics](https://torchmetrics.readthedocs.io/en/latest/) is here to solve.

Does this mean that if I had implemented my loss function (RMSE) as a torchmetric's `Metric` and just called `self.log` from `validation_step` the checkpoint callback would have worked fine without `sync_dist=True`?

**Comment by awaelchli** [comment]
Yes, and torchmetrics has it:

https://torchmetrics.readthedocs.io/en/latest/references/modules.html#meansquarederror

just set `squared=False` and it returns the RSME.

**Comment by jamarju** [comment]
> The warning is a nice idea, however note most of the time we don't want to log with sync-dist and such a warning could become annoying.

Agreed. I would not warn about logging non synced metrics. But I would certainly warn when trying to checkpoint based on a non-synced metric.

> Yes, and torchmetrics has it:

That's great. But there should be a pointer in that direction in my humble opinion.

**Comment by carmocca** [comment]
@awaelchli can we close this? Is there any reason why you have kept it open?

**Comment by awaelchli** [comment]
Kept it open for discussion if someone has ideas. 

I don't think we can warn in a meaningful way that is not intrusive and leads to false positives. It's a trade-off between accuracy (model checkpointing based on synced will be more accurate) and speed (syncing with a collective call), in the case described, we can't have both, and therefore a warning on either side would potentially be deceptive.
